core/src/main/java/org/elasticsearch/script/ScriptService.java

```diff
@@ -225,6 +248,31 @@ public CompiledScript compile(Script script, ScriptContext scriptContext, Map<St
     }
 
     /**
+     * Check whether there have been too many compilations within the last minute, throwing a circuit breaking exception if so.
+     */
+    public void checkCompilationLimit() {
+        long now = System.nanoTime();
```

Comment:
It'd be useful to describe the how here I think. I guess this works like a bucket of water slowly filling up. You get to compile a script if there is enough water in the bucket, otherwise you don't. If the bucket overflows then you don't get more water.